### DISCUSSION TITLE
Error when upgrading Python dependencies with hashes

### DISCUSSION THREAD

**Discussion by debfx** [renovatebot/renovate#32163] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab,  v38.132.1

### Please tell us more about your question or problem

Python requirements.in / requirements.txt dependency files that contain hashes fail to update properly.

The versions are updated but not the hashes, leading to a broken requirements.txt

### Logs (if relevant)

<details><summary>Logs</summary>

```

[04:09:23.386] INFO (509): Installing pip hashin@1.0.2...
[04:09:28.900] ERROR (509): Command failed with exit code 1: hashin --version
[04:09:28.901] FATAL (509): Install tool hashin failed in 5.5s.

Traceback (most recent call last):
  File "/opt/containerbase/tools/hashin/1.0.2/3.13.0/bin/hashin", line 5, in <module>
    from hashin import main
  File "/opt/containerbase/tools/hashin/1.0.2/3.13.0/lib/python3.13/site-packages/hashin.py", line 10, in <module>
    import cgi
ModuleNotFoundError: No module named 'cgi'

```

</details>

**Comment by rarkins** [comment]
Do you think it would be solved by using a different version of python?

**Reply by debfx** [reply]
Yes, hashin seems to be incompatible with Python >= 3.13 right now. It uses the cgi module that has been removed. https://github.com/peterbe/hashin/issues/160

Downgrading to Python 3.12 for now should fix the issue.

**Reply by rarkins** [reply]
You can configure "constraints" in renovate to control the version of python

**Reply by debfx** [reply]
Thanks, setting it to a lower version indeed works fine.

The fix for hashin was just merged so hopefully we'll get a new release compatible with 3.13 soon.
